src/deploy/functions/validate.ts

```diff
@@ -0,0 +1,87 @@
+import * as FirebaseError from "../../error";
+import * as _ from "lodash";
+import * as path from "path";
+import * as clc from "cli-color";
+import * as logger from "../../logger";
+import * as projectPath from "../../projectPath";
+import * as fsutils from "../../fsutils";
+
+// have to require this because no @types/cjson available
+// tslint:disable-next-line
+const cjson = require("cjson");
+
+/**
+ * Check that functions directory exists.
+ * @param cwd Working directory.
+ * @param sourceDirName Relative path to source directory.
+ * @throws { FirebaseError } Functions directory must exist.
+ */
+export function functionsDirectoryExists(cwd: string, sourceDirName: string): void {
+  if (!fsutils.dirExistsSync(projectPath.resolveProjectPath(cwd, sourceDirName))) {
+    const msg =
+      `could not deploy functions because the ${clc.bold('"' + sourceDirName + '"')} ` +
+      `directory was not found. Please create it or specify a different source directory in firebase.json`;
+    throw new FirebaseError(msg);
+  }
+}
+
+/**
+ * Validate function names only contain lower case letters, numbers, and dashes.
+ * @param functionNames Object containing function names as keys.
+ * @throws { FirebaseError } Function names must be valid.
+ */
+export function functionNamesAreValid(functionNames: {}): void {
+  const validFunctionNameRegex = /^[a-z][a-zA-Z0-9_-]{1,62}$/i;
```

Comment:
This seems like an unusual way to enforce the a functions name is all lowercase. Why do we match A-Z and then add a /i modifier to make it case insensitive?

Reply:
Fair question - this code was originally on the critical path in prepare. When I refactored it out, I got curious as to what it did, but turns out it's not actually being called (I also tried this on master branch without my refactor to ensure it's not from my changes). So, if you put in a bad string as your function name, it will fail further down in the API call, not in the validate step. Basically this is a roundabout way of saying I'm not sure where this gets the regex from - I tried to search in the cloud docs for a valid function name and wasn't able to find anything. If we can find the docs defining a valid function name, we can at least change the regex here to be correct. I have a bug out for fixing the part about this function not being called (131331234).

Reply:
Ah found it: https://cloud.google.com/functions/docs/deploying/filesystem#deploy_using_the_gcloud_tool

"This name can only contain letters, numbers, underscores, and hyphens"